### PR TITLE
Makes it possible to store abstract Entities in the index

### PR DESCRIPTION
**Pull request by D-02022020**
This makes it possible to store different subclass of an Entity class.
To do so, the parent Entity class must be abstract and have an Indexed Annotation:
```
@Indexed(index = "core",
        routing = ParentClass.ROUTING)
public abstract class ParentClass extends Entity {

    public abstract void abstractMethod();
    ...
```

The subclasses need to look like this:
```
@Indexed(index = "core",
        routing = ParentClass.ROUTING,
        subClassCode = SubClass1.SUBCLASS_CODE)
public class SubClass1 extends ParentClass {

    public static final String SUBCLASS_CODE = "code-for-this-specific-subclass";

    public abstract void abstractMethod() {
        ...
    }

    ...
```
```
@Indexed(index = "core",
        routing = ParentClass.ROUTING,
        subClassCode = SubClass2.SUBCLASS_CODE)
public class SubClass2 extends ParentClass {
    ...
```

Storing subclass instances is easy: `Index.update(subClassInstance)`.
This will store the subclass-code of this specific instance among the other fields in the index.

Retrieving subclass instances is easy, too, if you already know the type of the Entity:
- `SubClass instance = Index.find(routing, SubClass.class, id);`
- `List<SubClass> instances = Index.select(SubClass.class)....queryList();`

Otherwise you need to select on the ParentClass: 
- `ParentClass instance = Index.find(routing, ParentClass.class, id);`
- `List<ParentClass> instances = Index.select(ParentClass.class)....queryList();`
These instances are statically typed with the parent class, but are instances of the specific subclasses, each depending on the subclass-code that was stored in the index document. Calling Methods on this instance(s) will therefore call the correct subclass implementation due to Java's [dynamic method dispatching](http://www.studytonight.com/java/dynamic-method-dispatch.php)

When loading datasets via JSON, use the type of the subclass as `_type`:
```
{
    "_id": "ID",
    "_type": "core-SubClass1",
    ...
},
```

### CODE REVIEW

**Comment by D-02022020** [comment]
The unchecked casts in `Schema.addAbstractParentClass()` can never throw a `ClassCastException` except if the `Entity` class itself would have an `Indexed`-Annotation with filled `subClassCode`
The unchecked cast in `Index.createEntity` can never throw a `ClassCastException` as long as  `Schema.addAbstractParentClass()` does its job right (e.g. only stores Descriptors of subclasses in `parentDescriptor.getSubClassDescriptors().put(...)`)

File: src/main/java/sirius/search/EntityDescriptor.java [hunk 1]
```diff
@@ -223,10 +227,20 @@ public String toJson(Entity entity, @Nullable String objectName) {
     }
 
     /**
+     * Returns the class object of this descriptors {@link Entity} subclass
+     *
+     * @return the class object of this descriptors {@link Entity} subclass
+     */
+    public Class<? extends Entity> getClazz() {
```
Comment (andyHa) [inline src/main/java/sirius/search/EntityDescriptor.java]: getEntityType() ?

File: src/main/java/sirius/search/EntityDescriptor.java [hunk 2]
```diff
@@ -223,10 +227,20 @@ public String toJson(Entity entity, @Nullable String objectName) {
...
+        return clazz;
+    }
+
+    /**
      * Returns the name of the index which is used to store the data.
      * <p>
      * Note that this is an ElasticSearch index and not to be confused with a database index. This would be more or
      * less something like a schema in a SQL db
+     * </p>
```
Comment (andyHa) [inline src/main/java/sirius/search/EntityDescriptor.java]: nope - javadoc for j8 says just p tags

File: src/main/java/sirius/search/EntityDescriptor.java [hunk 3]
```diff
@@ -322,4 +366,23 @@ public String getRouting() {
...
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        EntityDescriptor that = (EntityDescriptor) o;
+        return java.util.Objects.equals(indexName, that.indexName)
```
Comment (andyHa) [inline src/main/java/sirius/search/EntityDescriptor.java]: import?

File: src/main/java/sirius/search/Index.java [hunk 4]
```diff
@@ -1543,4 +1573,53 @@ public static void loadDataset(String dataset) {
...
+                                                    @Nonnull String id,
+                                                    long version,
+                                                    @Nonnull Map<String, Object> source)
+            throws IllegalAccessException, InstantiationException {
+        EntityDescriptor descriptor = getDescriptor(clazz);
+        if (Modifier.isAbstract(clazz.getModifiers())) {
+            String subClassCode = (String) source.get(SUBCLASSCODE_FIELD);
+            EntityDescriptor subclassDescriptor = descriptor.getSubClassDescriptors().get(subClassCode);
+            if (subclassDescriptor == null) {
+                LOG.WARN("INSTANTIATE SUBCLASS: %s: no descriptor found for subClassCode \"%s\"!",
```
Comment (andyHa) [inline src/main/java/sirius/search/Index.java]: shouldn't we throw an exception here?

File: src/main/java/sirius/search/Index.java [hunk 5]
```diff
@@ -1543,4 +1573,53 @@ public static void loadDataset(String dataset) {
...
     }
+
+    public static <E extends Entity> E createEntity(@Nonnull Class<E> clazz,
+                                                    @Nonnull String id,
+                                                    long version,
+                                                    @Nonnull Map<String, Object> source)
+            throws IllegalAccessException, InstantiationException {
+        EntityDescriptor descriptor = getDescriptor(clazz);
+        if (Modifier.isAbstract(clazz.getModifiers())) {
+            String subClassCode = (String) source.get(SUBCLASSCODE_FIELD);
```
Comment (andyHa) [inline src/main/java/sirius/search/Index.java]: extract block into method?

File: src/main/java/sirius/search/Index.java [hunk 6]
```diff
@@ -1543,4 +1573,53 @@ public static void loadDataset(String dataset) {
...
+            }
+            try {
+                E subEntity = (E) subclassDescriptor.getClazz().newInstance();
+                subEntity.initSourceTracing();
+                subEntity.setId(id);
+                subEntity.setVersion(version);
+                subclassDescriptor.readSource(subEntity, source);
+                return subEntity;
+            } catch (Exception e) {
+                LOG.WARN(
```
Comment (andyHa) [inline src/main/java/sirius/search/Index.java]: throw?

File: src/main/java/sirius/search/Schema.java [hunk 7]
```diff
@@ -53,26 +69,82 @@
...
+    /**
+     * Searches for the first abstract parent class in the class hierarchy with the same {@link Indexed} annotation and
+     * creates a {@link EntityDescriptor} for it.
+     *
+     * @param subClass           the subclass
+     * @param subClassDescriptor descriptor of the subclass
+     * @param <E>                type of the subclass
+     * @return whether a matching parent class could be found
+     */
+    private <E extends Entity> boolean addAbstractParentClass(Class<E> subClass, EntityDescriptor subClassDescriptor) {
```
Comment (andyHa) [inline src/main/java/sirius/search/Schema.java]: brainoverload - refactor into several methods

File: src/main/java/sirius/search/EntityDescriptor.java [hunk 8]
```diff
@@ -257,6 +313,15 @@ public XContentBuilder createMapping() throws IOException {
         for (Property p : getProperties()) {
             p.createMapping(builder);
         }
+        builder.startObject(Index.SUBCLASSCODE_FIELD);
```
Comment (andyHa) [inline src/main/java/sirius/search/EntityDescriptor.java]: move to a separate method?

File: src/main/java/sirius/search/Schema.java [hunk 9]
```diff
@@ -41,6 +52,11 @@
 public class Schema {
 
     /**
+     * Logger used by this framework
+     */
+    public static final Log LOG = Log.get("index");
```
Comment (andyHa) [inline src/main/java/sirius/search/Schema.java]: wasn't there a logger anywhere before?

**Pull request merged**